python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -1261,3 +1266,57 @@ def _numpy_style_arrays_to_arrow_arrays(
                 raise err
 
     return arrow_arrays
+
+
+def compute_row_group_size(schema: pa.Schema, target_size: int = TARGET_ROW_GROUP_BYTES) -> int:
+    """Compute approximate row group size for a given arrow schema.
+
+    Given a schema, this routine will compute the number of rows in a row group
+    that will yield persisted size on disk of the target size or smaller.  The
+    exact size on disk depends on the compression settings and ratios; typical
+    binary data tables will have around 15-20% compression with the pyarrow
+    default ``snappy`` compression.
+
+    Parameters
+    ----------
+    schema : `pyarrow.Schema`
+        Arrow table schema.
+    target_size : `int`, optional
+        The target size (in bytes).
+
+    Returns
+    -------
+    row_group_size : `int`
+        Number of rows per row group to hit the target size.
+    """
+    bit_width = 0
+
+    metadata = schema.metadata if schema.metadata is not None else {}
+
+    for name in schema.names:
+        t = schema.field(name).type
+
+        if t in (pa.string(), pa.binary()):
+            md_name = f"lsst::arrow::len::{name}"
+
+            if (encoded := md_name.encode("UTF-8")) in metadata:
+                # String/bytes length from header.
+                strlen = int(schema.metadata[encoded])
+            else:
+                # We don't know the string width, so guess something.
+                strlen = 10
+
+            # Assuming UTF-8 encoding, and very few wide characters.
+            t_width = 8 * strlen
+        elif isinstance(t, pa.FixedSizeListType):
+            t_width = t.list_size * t.value_type.bit_width
+        elif t == pa.null():
+            t_width = 0
+        else:
+            t_width = t.bit_width
+
+        bit_width += t_width
+
+    byte_width = bit_width // 8
```

Comment:
Round up?  Also, is there a pathological case where `bit_width < 8` (in particular, `== 0`)?

Reply:
Great point, I can check for that.  I'm not too worried about rounding, these are all approximate anyway.